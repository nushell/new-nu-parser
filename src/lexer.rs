use crate::compiler::{Span, Spanned};
use logos::{Lexer, Logos};

#[derive(Debug, Default, Copy, Clone, PartialEq)]
pub enum LexError {
    #[default]
    Generic,
    UnmatchedStrInterpLParen,
    UnmatchedStrInterpRParen,
}

/// Average number of bytes per token used for estimating the tokens buffer size.
///
/// Estimated with this snippet:
/// let res = ls tests/**/*.nu | get name | each {|name|
///     let nbytes = open --raw $name | into binary | length
///     let ntokens = cargo run -- $name | lines | where $it starts-with 'Token3' | length
///     {
///         file: $name
///         nbytes: $nbytes
///         ntokens: $ntokens
///         bytes_per_token: ($nbytes / $ntokens)
///     }
/// }
///
/// TODO: Use larger and more representative codebase to estimate this
const AVG_BYTES_PER_TOKEN: usize = 2;

/// Lexed tokens
///
/// Tokens and spans are stored in separate vectors indexed by a position index (starting at 0).
#[derive(Clone)]
pub struct Tokens {
    pos: usize,
    tokens: Vec<Token>,
    spans: Vec<Span>,
}

impl Tokens {
    /// Create a new Tokens with allocated storage for the tokens and spans
    pub fn new(source: &[u8]) -> Self {
        let estimated_num_tokens = source.len() / AVG_BYTES_PER_TOKEN;
        Tokens {
            pos: 0,
            tokens: Vec::with_capacity(estimated_num_tokens),
            spans: Vec::with_capacity(estimated_num_tokens),
        }
    }

    // Position-related methods

    /// Advance position to point at the next token
    ///
    /// Note that this can potentially point beyond the tokens if called enough times. The parser
    /// should correctly check for EOF and terminate without advancing further.
    pub fn advance(&mut self) {
        debug_assert!(self.pos < self.tokens.len());
        self.pos += 1;
    }

    /// Return current position
    pub fn pos(&self) -> usize {
        self.pos
    }

    /// Set current position
    pub fn set_pos(&mut self, pos: usize) {
        self.pos = pos;
    }

    // Adding and fetching tokens

    /// Push a spanned token to the internal storage
    pub fn push(&mut self, token: Token, span: Span) {
        self.tokens.push(token);
        self.spans.push(span);
    }

    /// Check the token at the current position
    pub fn peek(&self) -> (Token, Span) {
        (self.peek_token(), self.peek_span())
    }

    /// Same as peek() but return only the token
    pub fn peek_token(&self) -> Token {
        self.tokens[self.pos]
    }

    /// Same as peek() but return only the span
    pub fn peek_span(&self) -> Span {
        self.spans[self.pos]
    }

    // Printing

    /// Format the tokens into a human-readable output for debugging
    pub fn display(&self, source: &[u8]) -> String {
        let mut result = String::new();

        result.push_str("==== TOKENS ====\n");

        for (i, (token, span)) in self.tokens.iter().zip(self.spans.iter()).enumerate() {
            result.push_str(&format!(
                "Token3 {i:4}: {:25} span: {:4} .. {:4} '{}'\n",
                format!("{:?}", token),
                span.start,
                span.end,
                String::from_utf8_lossy(
                    source
                        .get(span.start..span.end)
                        .expect("missing source of token span")
                )
                .replace("\r", "\\r")
                .replace("\n", "\\n")
                .replace("\t", "\\t")
            ));
        }

        result
    }

    /// Print the output of display() to standard output
    pub fn print(&self, source: &[u8]) {
        let output = self.display(source);
        print!("{output}");
    }

    /// Print the output of display() to standard error
    pub fn eprint(&self, source: &[u8]) {
        let output = self.display(source);
        eprint!("{output}");
    }
}

// TODO: Deduplicate code between lex_internal_dq_string_interp() and lex_internal_sq_string_interp()
/// Lex the contents of a double-quoted string interpolation
fn lex_internal_dq_string_interp(
    contents: &[u8],
    span_offset: usize,
    tokens: &mut Tokens,
) -> Result<(), Spanned<LexError>> {
    let lexer = DqStrInterpToken::lexer(contents).spanned();

    for (res, span) in lexer {
        let new_span = Span::new(span.start + span_offset, span.end + span_offset);
        match res {
            Ok(DqStrInterpToken::Start) => {
                tokens.push(Token::DqStringInterpStart, new_span);
            }
            Ok(DqStrInterpToken::StringChunk) => {
                tokens.push(Token::StrInterpChunk, new_span);
            }
            Ok(DqStrInterpToken::Subexpression) => {
                tokens.push(
                    Token::StrInterpLParen,
                    Span::new(new_span.start, new_span.start + 1),
                );

                lex_internal(
                    &contents[span.start + 1..span.end - 1],
                    span_offset + span.start + 1,
                    tokens,
                )?;

                tokens.push(
                    Token::StrInterpRParen,
                    Span::new(new_span.end - 1, new_span.end),
                );
            }
            Ok(DqStrInterpToken::End) => {
                tokens.push(Token::StrInterpEnd, new_span);
                return Ok(());
            }
            Err(e) => {
                return Err(Spanned::new(e, new_span));
            }
        }
    }

    Ok(())
}

// TODO: Deduplicate code between lex_internal_dq_string_interp() and lex_internal_sq_string_interp()
/// Lex the contents of a single-quoted string interpolation
fn lex_internal_sq_string_interp(
    contents: &[u8],
    span_offset: usize,
    tokens: &mut Tokens,
) -> Result<(), Spanned<LexError>> {
    let lexer = SqStrInterpToken::lexer(contents).spanned();

    for (res, span) in lexer {
        let new_span = Span::new(span.start + span_offset, span.end + span_offset);
        match res {
            Ok(SqStrInterpToken::Start) => {
                tokens.push(Token::SqStringInterpStart, new_span);
            }
            Ok(SqStrInterpToken::StringChunk) => {
                tokens.push(Token::StrInterpChunk, new_span);
            }
            Ok(SqStrInterpToken::Subexpression) => {
                tokens.push(
                    Token::StrInterpLParen,
                    Span::new(new_span.start, new_span.start + 1),
                );

                lex_internal(
                    &contents[span.start + 1..span.end - 1],
                    span_offset + span.start + 1,
                    tokens,
                )?;

                tokens.push(
                    Token::StrInterpRParen,
                    Span::new(new_span.end - 1, new_span.end),
                );
            }
            Ok(SqStrInterpToken::End) => {
                tokens.push(Token::StrInterpEnd, new_span);
                return Ok(());
            }
            Err(e) => {
                return Err(Spanned::new(e, new_span));
            }
        }
    }

    Ok(())
}

fn lex_internal(
    contents: &[u8],
    span_offset: usize,
    tokens: &mut Tokens,
) -> Result<(), Spanned<LexError>> {
    let lexer = Token::lexer(contents).spanned();

    for (res, span) in lexer {
        let new_span = Span::new(span.start + span_offset, span.end + span_offset);
        match res {
            Ok(Token::DqStrInterp) => lex_internal_dq_string_interp(
                &contents[span.start..span.end],
                span_offset + span.start,
                tokens,
            )?,
            Ok(Token::SqStrInterp) => lex_internal_sq_string_interp(
                &contents[span.start..span.end],
                span_offset + span.start,
                tokens,
            )?,
            Ok(token) => tokens.push(token, new_span),
            Err(e) => {
                return Err(Spanned::new(e, new_span));
            }
        }
    }

    Ok(())
}

/// Lex the source contents and return allocated Tokens.
///
/// In the case of error, you can look up the last stored token to get a clue what went wrong. The
/// last stored token is always End Of File (EOF), so there will always be at least one token.
pub fn lex(contents: &[u8], span_offset: usize) -> (Tokens, Result<(), Spanned<LexError>>) {
    // TODO: We might require the contents to always end with a newline, in which case return an error
    let mut tokens = Tokens::new(contents);
    let res = lex_internal(contents, span_offset, &mut tokens);

    tokens.push(
        Token::Eof,
        Span::new(contents.len() + span_offset, contents.len() + span_offset),
    );

    if let Err(e) = res {
        return (tokens, Err(e));
    }

    (tokens, Ok(()))
}

#[derive(Logos, Debug, Clone, Copy, PartialEq)]
#[logos(skip r"[ \t]+")]
#[logos(source = [u8], error = LexError)]
pub enum Token {
    #[regex("(0[xob])?[0-9][0-9_]*", priority = 10)]
    Int,
    #[regex(r"([0-9][0-9_]*)*\.([0-9][0-9_]*)*([eE][+-]?[0-9_]+)?")]
    Float,
    #[regex("\n|\r\n|\x0C")]
    Newline,
    #[regex(r#""([^"\\]|\\["\\bnfrt])*""#)]
    DoubleQuotedString,
    #[regex(r#"'[^']*'"#)]
    SingleQuotedString,
    #[regex(r#"`[^`]*`"#)]
    BacktickBareword,
    // #[regex(r#"[ \t]+"#)]
    // HorizontalWhitespace,
    #[regex(r#"[0-9]{4}-[0-9]{2}-[0-9]{2}(T[0-9]{2}:[0-9]{2}:[0-9]{2}(\.[0-9]+)?)?(Z|[\+-][0-9]{2}:[0-9]{2})?"#)]
    Datetime,
    #[regex(r#"#[^\n]*"#, priority = 20)]
    Comment,
    // lower priority to avoid clashing with Int
    #[regex(r#"(_|[^\s[:punct:]])(#|_|[^\s[:punct:]])*"#, priority = 2)]
    Bareword,
    #[token("...")]
    DotDotDot,
    #[token("..")]
    DotDot,
    #[token(".", priority = 10)] // higher priority to avoid "." being tokenized as Float
    Dot,
    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("[")]
    LSquare,
    #[token("]")]
    RSquare,
    #[token("{")]
    LCurly,
    #[token("}")]
    RCurly,
    #[token("<=")]
    LessThanEqual,
    #[token("<")]
    LessThan,
    #[token(">=")]
    GreaterThanEqual,
    #[token(">")]
    GreaterThan,
    #[token("++")]
    PlusPlus,
    #[token("+=")]
    PlusEquals,
    #[token("+")]
    Plus,
    #[token("->")]
    ThinArrow,
    #[token("=>")]
    ThickArrow,
    #[token("-=")]
    DashEquals,
    #[token("-")]
    Dash,
    #[token("**")]
    AsteriskAsterisk,
    #[token("*=")]
    AsteriskEquals,
    #[token("*")]
    Asterisk,
    #[token("//")]
    ForwardSlashForwardSlash,
    #[token("/=")]
    ForwardSlashEquals,
    #[token("/")]
    ForwardSlash,
    #[token("==")]
    EqualsEquals,
    #[token("=~")]
    EqualsTilde,
    #[token("=")]
    Equals,
    #[token("::")]
    ColonColon,
    #[token(":")]
    Colon,
    #[token("$")]
    Dollar,
    #[token(";")]
    Semicolon,
    #[token("!=")]
    ExclamationEquals,
    #[token("!~")]
    ExclamationTilde,
    #[token("!")]
    Exclamation,
    #[token("&&")]
    AmpersandAmpersand,
    #[token("&")]
    Ampersand,
    #[token(",")]
    Comma,
    #[token("?")]
    QuestionMark,
    #[token("^")]
    Caret,
    #[token("@")]
    At,
    #[token("||")]
    PipePipe,
    #[token("|")]
    Pipe,
    #[token("o>")]
    OutGreaterThan,
    #[token("o>>")]
    OutGreaterGreaterThan,
    #[token("e>")]
    ErrGreaterThan,
    #[token("e>>")]
    ErrGreaterGreaterThan,
    #[token("o+e>")]
    OutErrGreaterThan,
    #[token("o+e>>")]
    OutErrGreaterGreaterThan,
    #[token("e>|")]
    ErrGreaterThanPipe,
    #[token("o+e>|")]
    OutErrGreaterThanPipe,
    /// Double quoted string interpolation $"..."
    ///
    /// The token is passed to a separate lexer and is not actually present in the result.
    /// Unescaped double quotes are not permitted, for example, $"foo("bar")" is not allowed.
    #[regex(r#"\$"([^"]|\\")*""#)]
    DqStrInterp,
    /// Single-quoted string interpolation $'...'
    ///
    /// The token is passed to a separate lexer and is not actually present in the result.
    #[regex(r#"\$'[^']*'"#)]
    SqStrInterp,
    /// Start of double-quoted string interpoloation $" (returned from separate lexing)
    DqStringInterpStart,
    /// Start of single-quoted string interpoloation $' (returned from separate lexing)
    SqStringInterpStart,
    /// Non-interpolated string chunk within any string interpolation (returned from separate lexing)
    ///
    /// For example, "foo" within $"foo(1)"
    StrInterpChunk,
    /// Left parenthesis inside any string interpolation (returned from separate lexing)
    StrInterpLParen,
    /// Right parenthesis inside any string interpolation (returned from separate lexing)
    StrInterpRParen,
    /// End of any string interpolation (returned from separate lexing)
    StrInterpEnd,
    /// End of file, doesn't match any syntax, but lexed tokens always end with it
    Eof,
}

fn match_subexpression<'a, T: Logos<'a>>(
    remainder: &[u8],
    lexer: &mut Lexer<'a, T>,
) -> Result<(), LexError> {
    let mut depth = 1;
    let mut pos = 0;

    while pos < remainder.len() {
        match remainder[pos] {
            b'(' => depth += 1,
            b')' => depth -= 1,
            _ => (),
        }

        if depth == 0 {
            break;
        }

        if depth < 0 {
            // unmatched )
            return Err(LexError::UnmatchedStrInterpRParen);
        }

        pos += 1;
    }

    if depth > 0 {
        // unmatched (
        return Err(LexError::UnmatchedStrInterpLParen);
    }

    lexer.bump(pos + 1);
    Ok(())
}

/// Tokens representing double-quoted string interpolation
#[derive(Logos, Debug, Clone, Copy, PartialEq)]
#[logos(source = [u8], error = LexError)]
enum DqStrInterpToken {
    #[token(r#"$""#)]
    Start,
    #[regex(r#"([^"\\\(]|\\["\\bnfrt\(])+"#)]
    StringChunk,
    #[token("(", |lex| match_subexpression(lex.remainder(), lex))]
    Subexpression,
    #[token(r#"""#)]
    End,
}

/// Tokens representing single-quoted string interpolation
#[derive(Logos, Debug, Clone, Copy, PartialEq)]
#[logos(source = [u8], error=LexError)]
enum SqStrInterpToken {
    #[token(r#"$'"#)]
    Start,
    #[regex(r#"[^'\(]+"#)]
    StringChunk,
    #[token("(", |lex| match_subexpression(lex.remainder(), lex))]
    Subexpression,
    #[token(r#"'"#)]
    End,
}

#[cfg(test)]
mod test {
    /// Lexer tests useful for smaller sources, errors and corner cases
    use crate::compiler::{Span, Spanned};
    use crate::lexer::{lex, Token};

    use super::LexError;

    fn test_lex(
        src: &[u8],
        expected_tokens: &[(Token, Span)],
        expected_result: Result<(), Spanned<LexError>>,
    ) {
        let (mut actual_tokens, actual_result) = lex(src, 0);

        assert_eq!(expected_result, actual_result, "Lexing result mismatch");

        for (i, expected) in expected_tokens.iter().enumerate() {
            let actual = actual_tokens.peek();
            assert_eq!(expected, &actual, "Mismatch in token {}", i);
            actual_tokens.advance();
        }
    }

    fn span(start: usize, end: usize) -> Span {
        Span { start, end }
    }

    #[test]
    fn lex_last_eof() {
        test_lex(b"", &[(Token::Eof, span(0, 0))], Ok(()));
    }

    #[test]
    fn lex_unmatched_string() {
        // TODO: Make unmatched delimiters nicer
        test_lex(
            b"'unmatched string",
            &[(Token::Eof, span(17, 17))],
            Err(Spanned::new(LexError::Generic, Span::new(0, 17))),
        );
    }

    #[test]
    fn lex_string_interp_errors() {
        test_lex(
            br#"$"foo("baz")bar""#,
            &[
                (Token::DqStringInterpStart, span(0, 2)),
                (Token::StrInterpChunk, span(2, 5)),
                (Token::Eof, span(16, 16)),
            ],
            Err(Spanned::new(
                LexError::UnmatchedStrInterpLParen,
                Span::new(5, 6),
            )),
        );

        test_lex(
            br#"$'foo('baz')bar'"#,
            &[
                (Token::SqStringInterpStart, span(0, 2)),
                (Token::StrInterpChunk, span(2, 5)),
                (Token::Eof, span(16, 16)),
            ],
            Err(Spanned::new(
                LexError::UnmatchedStrInterpLParen,
                Span::new(5, 6),
            )),
        );
    }
}
