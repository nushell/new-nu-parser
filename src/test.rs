use crate::lexer::lex;
use crate::resolver::Resolver;
use crate::typechecker::Typechecker;
use crate::{compiler::Compiler, parser::Parser};

use std::path::Path;

fn evaluate_example(fname: &Path) -> String {
    let mut compiler = Compiler::new();
    let contents = std::fs::read(fname).expect("We only run tests found by glob");

    let span_offset = compiler.span_offset();
    compiler.add_file(&fname.to_string_lossy(), &contents);

    let (tokens, err) = lex(&contents, span_offset);
    if let Err(e) = err {
        tokens.eprint(&contents);
        eprintln!("Lexing error. Error: {:?}", e);
        std::process::exit(1);
    }

    let parser = Parser::new(compiler, tokens);
    compiler = parser.parse();

    let mut result = compiler.display_state();

    if !compiler.errors.is_empty() {
        return result;
    }

    let mut resolver = Resolver::new(&compiler);
    resolver.resolve();
    result.push_str(&resolver.display_state());

    compiler.merge_name_bindings(resolver.to_name_bindings());

    if !compiler.errors.is_empty() {
        return result;
    }

    let mut typechecker = Typechecker::new(&compiler);
    typechecker.typecheck();
    result.push_str(&typechecker.display_state());

    compiler.merge_types(typechecker.to_types());

    result
}

fn evaluate_lexer(fname: &Path) -> String {
    let contents = std::fs::read(fname);

    let Ok(contents) = contents else {
        panic!("Lexer: can't find file {}", fname.to_string_lossy());
    };

    let (tokens, err) = lex(&contents, 0);
    let mut res = tokens.display(&contents);

    if let Err(e) = err {
        res.push_str(&format!("Lexing error. Error: {:?}", e));
    }

    res
}

#[test]
fn test_node_output() {
    insta::glob!("../tests", "*.nu", |path| {
        insta::assert_snapshot!(evaluate_example(path));
    });
}

#[test]
fn test_lexer() {
    insta::glob!("../tests/lex", "*.nu", |path| {
        insta::assert_snapshot!(evaluate_lexer(path));
    });
}
