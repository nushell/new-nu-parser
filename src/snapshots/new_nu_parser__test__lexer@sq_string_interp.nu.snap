---
source: src/test.rs
expression: evaluate_lexer(path)
input_file: tests/lex/sq_string_interp.nu
---
==== TOKENS ====
Token3    0: SqStringInterpStart       span:    0 ..    2 '$''
Token3    1: StrInterpEnd              span:    2 ..    3 '''
Token3    2: Newline                   span:    3 ..    4 '\n'
Token3    3: SqStringInterpStart       span:    4 ..    6 '$''
Token3    4: StrInterpChunk            span:    6 ..    9 'foo'
Token3    5: StrInterpEnd              span:    9 ..   10 '''
Token3    6: Newline                   span:   10 ..   11 '\n'
Token3    7: SqStringInterpStart       span:   11 ..   13 '$''
Token3    8: StrInterpChunk            span:   13 ..   17 'foo)'
Token3    9: StrInterpEnd              span:   17 ..   18 '''
Token3   10: Newline                   span:   18 ..   19 '\n'
Token3   11: SqStringInterpStart       span:   19 ..   21 '$''
Token3   12: StrInterpLParen           span:   21 ..   22 '('
Token3   13: Int                       span:   22 ..   23 '1'
Token3   14: StrInterpRParen           span:   23 ..   24 ')'
Token3   15: StrInterpChunk            span:   24 ..   27 'bar'
Token3   16: StrInterpEnd              span:   27 ..   28 '''
Token3   17: Newline                   span:   28 ..   29 '\n'
Token3   18: SqStringInterpStart       span:   29 ..   31 '$''
Token3   19: StrInterpChunk            span:   31 ..   34 'foo'
Token3   20: StrInterpLParen           span:   34 ..   35 '('
Token3   21: Int                       span:   35 ..   36 '1'
Token3   22: StrInterpRParen           span:   36 ..   37 ')'
Token3   23: StrInterpChunk            span:   37 ..   40 'bar'
Token3   24: StrInterpEnd              span:   40 ..   41 '''
Token3   25: Newline                   span:   41 ..   42 '\n'
Token3   26: SqStringInterpStart       span:   42 ..   44 '$''
Token3   27: StrInterpChunk            span:   44 ..   47 'foo'
Token3   28: StrInterpLParen           span:   47 ..   48 '('
Token3   29: LParen                    span:   48 ..   49 '('
Token3   30: Int                       span:   49 ..   50 '1'
Token3   31: RParen                    span:   50 ..   51 ')'
Token3   32: StrInterpRParen           span:   51 ..   52 ')'
Token3   33: StrInterpChunk            span:   52 ..   55 'bar'
Token3   34: StrInterpEnd              span:   55 ..   56 '''
Token3   35: Newline                   span:   56 ..   57 '\n'
Token3   36: SqStringInterpStart       span:   57 ..   59 '$''
Token3   37: StrInterpChunk            span:   59 ..   62 'foo'
Token3   38: StrInterpLParen           span:   62 ..   63 '('
Token3   39: Int                       span:   63 ..   64 '1'
Token3   40: Plus                      span:   65 ..   66 '+'
Token3   41: LParen                    span:   67 ..   68 '('
Token3   42: Int                       span:   68 ..   69 '3'
Token3   43: Plus                      span:   70 ..   71 '+'
Token3   44: Int                       span:   72 ..   73 '4'
Token3   45: RParen                    span:   73 ..   74 ')'
Token3   46: StrInterpRParen           span:   74 ..   75 ')'
Token3   47: StrInterpChunk            span:   75 ..   78 'bar'
Token3   48: StrInterpEnd              span:   78 ..   79 '''
Token3   49: Newline                   span:   79 ..   80 '\n'
Token3   50: SqStringInterpStart       span:   80 ..   82 '$''
Token3   51: StrInterpChunk            span:   82 ..   85 'foo'
Token3   52: StrInterpLParen           span:   85 ..   86 '('
Token3   53: StrInterpRParen           span:   86 ..   87 ')'
Token3   54: StrInterpChunk            span:   87 ..   90 'bar'
Token3   55: StrInterpEnd              span:   90 ..   91 '''
Token3   56: Newline                   span:   91 ..   92 '\n'
Token3   57: SqStringInterpStart       span:   92 ..   94 '$''
Token3   58: StrInterpChunk            span:   94 ..   97 'foo'
Token3   59: StrInterpLParen           span:   97 ..   98 '('
Token3   60: DqStringInterpStart       span:   98 ..  100 '$"'
Token3   61: StrInterpLParen           span:  100 ..  101 '('
Token3   62: Int                       span:  101 ..  102 '1'
Token3   63: Plus                      span:  103 ..  104 '+'
Token3   64: LParen                    span:  105 ..  106 '('
Token3   65: Int                       span:  106 ..  107 '2'
Token3   66: Plus                      span:  108 ..  109 '+'
Token3   67: Int                       span:  110 ..  111 '3'
Token3   68: RParen                    span:  111 ..  112 ')'
Token3   69: StrInterpRParen           span:  112 ..  113 ')'
Token3   70: StrInterpEnd              span:  113 ..  114 '"'
Token3   71: StrInterpRParen           span:  114 ..  115 ')'
Token3   72: StrInterpChunk            span:  115 ..  118 'bar'
Token3   73: StrInterpEnd              span:  118 ..  119 '''
Token3   74: Newline                   span:  119 ..  120 '\n'
Token3   75: Eof                       span:  120 ..  120 ''
